### PR TITLE
[READY] - Added Changelog Generation Script

### PR DESCRIPTION
**Pull request by maishiroma**
# What This PR Is
This PR adds in the following:
- A new python library, gen_release_notes, that helps automate the creation of release notes for a repository
- unit testing to verify the code and logic in said library
- documentation on how to use, configure and install said library
- a sample configuration file

## Tested
- [x] tox
- [x] installed and ran locally

### CODE REVIEW

File: scripts/gen_release_notes/gen_release_sample.config [hunk 1]
```diff
@@ -0,0 +1,28 @@
+ {
+    "tag_regex": "^[0-9]{1,}.[0-9]{1,}.[0-9]{1,}$",
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/gen_release_sample.config:2]: Thanks for making this configurable! This is pretty slick!

File: scripts/gen_release_notes/gen_release_sample.config [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    "rc_regex": "^RC/[0-9]{1,}.[0-9]{1,}$",
+    "bug_fix_regex": [
+        "fix",
+        "revert",
+        "patch"
+    ],
+    "main_branch_name": "master",
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/gen_release_sample.config:9]: good call checking for which branch is default

File: scripts/gen_release_notes/generate.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        json_obj = json.load(config_file)
+except FileNotFoundError:
+    print("Error, there doesn't exist a configuration file at {path}".format(path=file_path))
+    print("Place a config file in said path or specify a custom path via setting CONFIG_LOC.")
+    print("A sample config can be obtained in Nebulaworks/orion/scripts/gen_release_notes/gen_release_sample.config")
+    sys.exit(1)
+
+main_branch_name = json_obj["main_branch_name"]
+dev_env_name = json_obj["dev_env_name"]
+tst_env_name = json_obj["tst_env_name"]
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/generate.py:34]: can we go call this `stg` for `staging` instead? Only reason I say that is calling this `tst` is somewhat confusing and misleading.

File: scripts/gen_release_notes/tests/test_git_parser.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    result = git_parser.get_repo_information(mocked_repo)
+
+    assert result is not None
+    assert (
+        result["owner_name"] == user_input_good_remote_urls["owner"]
+        and result["repo_name"] == user_input_good_remote_urls["repo"]
+    )
+
+
+@pytest.mark.parametrize("user_input_bad_remote_urls", sample_bad_remote_urls)
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_git_parser.py:114]: Ooo parametrize 🙂

File: scripts/gen_release_notes/tests/test_git_parser.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    )
+
+
+@pytest.mark.parametrize("user_input_pr_information", sample_pr_information)
+def test_get_pr_information(user_input_pr_information):
+    """
+    Test to check if the information gathered in a PR is valid and the data structures are present
+    """
+
+    result = git_parser.get_pr_information(user_input_pr_information)
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_git_parser.py:140]: nice and simple test cases good job!

File: scripts/gen_release_notes/tests/test_git_parser.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    with patch.object(git_parser, "get_all_changed_files", return_value=[{"filename": "somefile"}]):
+        result = git_parser.get_pr_change_type(
+            user_input_pr_type["repo_owner"],
+            user_input_pr_type["repo_name"],
+            user_input_pr_type["pr_number"],
+            user_input_pr_type["pr_commit_message"],
+        )
+
+        assert result is not None
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_git_parser.py:160]: I think this none assert is redundant with your next assert.

File: scripts/gen_release_notes/tests/test_git_parser.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+@pytest.mark.parametrize("user_input_pr_information", sample_pr_information)
+def test_get_pr_information(user_input_pr_information):
+    """
+    Test to check if the information gathered in a PR is valid and the data structures are present
+    """
+
+    result = git_parser.get_pr_information(user_input_pr_information)
+
+    assert result is not None
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_git_parser.py:142]: Redundant none assert here too

File: scripts/gen_release_notes/tests/test_file_writer.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    Test to check if the file generation works as intended.
+
+    Asserts if we created a file that has contents in it
+    """
+
+    mock_env = Mock()
+    mock_env.next_release_tag_branch = Mock(return_value=user_input_cli_args["next_release_tag_branch"])
+    mock_env.sprint_number = Mock(return_value=user_input_cli_args["sprint_number"])
+
+    file_writer.write_contents_to_file(user_input_changelog_obj, user_input_dir_list, mock_env, output_path)
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_file_writer.py:57]: I would prefer we mock the writing of the object to disk so that way we dont have to call `os.remove`. @davido-nw might have some good tips for doing this but if we want to do this later im ok with that too.

File: scripts/gen_release_notes/README.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+| dev\_env\_name | Name of development environment. | `"dev"` |
+| misc\_env\_name | Name of general/non-environment specific changes. | `"general"` |
+| root\_env\_dir\_name | Name of the directory that corresponds to the start of the environment roots | `"env"` |
+| repo\_url | Repository to analyze with this script. | `"https://github.com/Nebulaworks/orion"` |
+| url\_repo\_endpoint | URL endpoint that corresponds to what endpoint said `repo_url` corresponds towards. | `"https://github.com"` |
+| pr\_body\_prefix | String to determine what pattern to look for in pull request titles. | `"] -"` |
+| output\_file\_name | Path to output the results. | `"gen_output_results.md"`|
+| weight\_scale | Scale to determine change priority. More details [below](#weight-scale). | `["env": "1", "modules": "10", "docs": "20"]` |
+
+### Weight Scale
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/README.md:173]: Thanks for explaining the scale I needed a refresher for how this worked myself 😅

File: scripts/gen_release_notes/README.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+| rc\_regex | String value to check for release candidate (e.g. specific branches) in a repo. Note that special characters (i.e. `.`) do not need escape characters. | `"^RC/[0-9]{1,}.[0-9]{1,}$"` | =
+| bug\_fix\_regex | List of values that correspond to a bug fix. | `["fix","hotfix","bug"]` |
+| main\_branch\_name | Name of the main/master branch in a repo. | `"master"` |
+| tst\_env\_name | Name of test environment. | `"tst"` |
+| prd\_env\_name | Name of production environment. | `"prd"` |
+| dev\_env\_name | Name of development environment. | `"dev"` |
+| misc\_env\_name | Name of general/non-environment specific changes. | `"general"` |
+| root\_env\_dir\_name | Name of the directory that corresponds to the start of the environment roots | `"env"` |
+| repo\_url | Repository to analyze with this script. | `"https://github.com/Nebulaworks/orion"` |
+| url\_repo\_endpoint | URL endpoint that corresponds to what endpoint said `repo_url` corresponds towards. | `"https://github.com"` |
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/README.md:168]: Is `url_repo_endpoint` always just the domain name of `repo_url`? Im curious what use case would cause them to diverge?

File: scripts/gen_release_notes/tests/test_helper.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+@pytest.mark.parametrize("user_regex_match", sample_regex_match)
+def test_is_regex_match(user_regex_match):
+    """
+    Test to confirm the regex match works as intended
+    """
+
+    results = helper.is_regex_match(user_regex_match["regex"], user_regex_match["context"])
+
+    assert results is not None and (results is False or results is True)
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_helper.py:99]: This will always pass no matter the match 😆  I think you want to add the desired outcome for the assert in the parametrize inputs.

File: scripts/gen_release_notes/tests/test_helper.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+@pytest.mark.parametrize("user_regex_search", sample_regex_search)
+def test_is_regex_search(user_regex_search):
+    """
+    Test to confirm the regex search works as expected
+    """
+
+    results = helper.is_regex_search(user_regex_search["regex"], user_regex_search["context"])
+
+    assert results is not None and (results is False or results is True)
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_helper.py:110]: Same here

File: scripts/gen_release_notes/tests/test_helper.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+@pytest.mark.parametrize("user_check_key_dict", sample_check_key_dict)
+def test_check_if_key_in_list_dict(user_check_key_dict):
+    """
+    Test to confirm that the search for a specific dict value works as intended
+    """
+
+    results = helper.check_if_key_in_list_dict(
+        user_check_key_dict["dicts"], user_check_key_dict["search_key"], user_check_key_dict["search_value"]
+    )
+    if results != -1:
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_helper.py:122]: Same here I think we can just add this to inputs of parametrize

**Review by sarcasticadmin — COMMENTED** [review]
@maishiroma awesome, thank you submitting this into `orion`! Overall I really like what youve got here!

Had a couple of small comments that Ive made inline and some things I wanted to call out overall:

1.  `assert result is not None` is called out in a few of the tests followed by another assert checking the same object for contents. I dont think we need both tests so double check to see if we can remove the none check.
2. Some of the tests cases should be checking specifically for the result (true or false) based on which item in the list/dict instead of doing `assert results is not None and (results is False or results is True)` 

Let me know what you think 🙂

File: scripts/gen_release_notes/generate.py (hunk 3 above)
Comment (maishiroma) [reply scripts/gen_release_notes/generate.py:34]: Oh good catch on that, I can refactor this script to use that name instead.

File: scripts/gen_release_notes/tests/test_git_parser.py (hunk 6 above)
Comment (maishiroma) [reply scripts/gen_release_notes/tests/test_git_parser.py:160]: Hm, I see what you mean here, originally this assert is to verify that one of the potential outputs of the function is NOT `None`, since that can be one of the potential outputs for this script. However, the following assert implies that results is already not `None`, so I can clean this out.

File: scripts/gen_release_notes/git_parser/git_parser.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    """
+    This function grabs all of the changed files that are in a given PR, given the following:
+    - repo owner
+    - repo name
+    - pr number
+
+    Returns a list of all changed files. If it cannot find any,
+    """
+
+    payload = {"Authorization": "token {secret}".format(secret=pat)}
```
Comment (davido-nw) [inline scripts/gen_release_notes/git_parser/git_parser.py:32]: Would it make more sense if the script just errors out if the `GITHUB_TOKEN` isn't defined (line 13 implies this)?

Alternatively - if it's possible to call this API without authentication (unlikely), then this probably shouldn't be part of the payload.

File: scripts/gen_release_notes/README.md (hunk 10 above)
Comment (maishiroma) [reply scripts/gen_release_notes/README.md:168]: That is correct; this is in the off chance that if someone has their repository in a different repository site (bitbucket, gitlab, github enterprise, etc)

File: scripts/gen_release_notes/git_parser/git_parser.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+                    owner=repo_owner, repo=repo_name, pr=pr_number, counter=iterator
+                ),
+                headers=payload,
+                timeout=2.0,
+            )
+        except requests.exceptions.RequestException:
+            print("Error getting files from PR. Did you pass in the right values? Or are you authenticated?")
+            return []
+
+        result = result + res.json()
```
Comment (davido-nw) [inline scripts/gen_release_notes/git_parser/git_parser.py:50]: Not urgent but FYI - this can be `result.extend(res.json())`.

File: scripts/gen_release_notes/tests/test_helper.py (hunk 11 above)
Comment (maishiroma) [reply scripts/gen_release_notes/tests/test_helper.py:99]: That was what I initially did actually. I was then reasoning out how the inputs probably shouldn't have logic that determines the outcome of a test. 

However, I think after letting this simmer a bit, I think the path I was thinking if the inputs correspond to some logic in the test that COULD configure a specific var in the test itself. Not the expected end result (if that makes sense).

Either way, I can change it back to what I did previously, it makes this easier to read

File: scripts/gen_release_notes/git_parser/git_parser.py [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+    - repo_name
+
+    Requires for there to be a remote named "origin" in the current local repo clone
+
+    Returns a dict with said values. If it cannot find both, returns an empty dict
+    """
+
+    url = repo_obj.remotes["origin"].url
+    colon_find = url.find(":")
+    dot_r_find = url.rfind(".")
```
Comment (davido-nw) [inline scripts/gen_release_notes/git_parser/git_parser.py:72]: 👍🏼 for `rfind`

File: scripts/gen_release_notes/README.md (hunk 10 above)
Comment (maishiroma) [reply scripts/gen_release_notes/README.md:168]: Hm actually, now that I think about this, I think I can just infer this value from `repo_endpoint`. Let me fix this up.

File: scripts/gen_release_notes/tests/test_file_writer.py (hunk 8 above)
Comment (davido-nw) [reply scripts/gen_release_notes/tests/test_file_writer.py:57]: This might be useful: https://docs.python.org/3/library/unittest.mock.html#mock-open

Comment (davido-nw) [reply scripts/gen_release_notes/tests/test_file_writer.py:57]: Also - check out the test cases I wrote for the AWS key rotator, since that's a script that involves writing to a file. In particular, this pattern where `mock_open` instantiates a "mock file" with data, then the patch to `builtins.open` makes all `open` syscalls use that mock file:
```py
fake_creds_fp = mock_open(read_data=creds_data)
with patch("builtins.open", fake_creds_fp):
    rotator.IAMKeyRotator(config).main()
````

Comment (sarcasticadmin) [reply scripts/gen_release_notes/tests/test_file_writer.py:57]: thanks @davido-nw for passing this along

File: scripts/gen_release_notes/git_parser/git_parser.py (hunk 14 above)
Comment (maishiroma) [reply scripts/gen_release_notes/git_parser/git_parser.py:32]: Good point; this script requires that value to be here, so it would make sense to just fail fast if it can't find that token.

File: scripts/gen_release_notes/git_parser/git_parser.py (hunk 15 above)
Comment (maishiroma) [reply scripts/gen_release_notes/git_parser/git_parser.py:50]: Ooo nice, thanks for the tip!

File: scripts/gen_release_notes/tests/test_file_writer.py (hunk 8 above)
Comment (maishiroma) [reply scripts/gen_release_notes/tests/test_file_writer.py]: Thanks for the help on this. I almost got it to work with this instead; I'm currently running into an issue with the `mock_open` also patching out one of the jinja2 methods, which is causing the following errors to occur:

```
________________________________________ test_write_to_file[user_input_cli_args1-user_input_dir_list1-user_input_changelog_obj0] ________________________________________

user_input_changelog_obj = {'dev': [{'change_entries': ['env/dev/someproject/something/nice.tf', 'docs/something/data.md', 'docs/something/anothe...'change_entries': ['awslambda/anotheroot/changes.py', 'awslamda/anothertest/changes.py'], 'type_change': 'awslambda'}]}
user_input_dir_list = ['dev', 'tst'], user_input_cli_args = {'next_release_tag_branch': 'another_branch', 'sprint_number': '31'}

    @pytest.mark.parametrize("user_input_changelog_obj", sample_changelog_list)
    @pytest.mark.parametrize("user_input_dir_list", sample_dir_list)
    @pytest.mark.parametrize("user_input_cli_args", sample_cli_args)
    def test_write_to_file(user_input_changelog_obj, user_input_dir_list, user_input_cli_args):
        """
        Test to check if the file generation works as intended.

        Asserts if we created a file that has contents in it
        """

        mock_env = Mock()
        mock_env.next_release_tag_branch = Mock(return_value=user_input_cli_args["next_release_tag_branch"])
        mock_env.sprint_number = Mock(return_value=user_input_cli_args["sprint_number"])

        with patch("builtins.open", mock_open()) as mocked_file:
>           file_writer.write_contents_to_file(user_input_changelog_obj, user_input_dir_list, mock_env, output_path)

tests/test_file_writer.py:57:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
file_writer/file_writer.py:23: in write_contents_to_file
    top_template = jinja_env.get_template("beginning.j2")
.tox/py38/lib/python3.8/site-packages/jinja2/environment.py:997: in get_template
    return self._load_template(name, globals)
.tox/py38/lib/python3.8/site-packages/jinja2/environment.py:958: in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
.tox/py38/lib/python3.8/site-packages/jinja2/loaders.py:125: in load
    source, filename, uptodate = self.get_source(environment, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <jinja2.loaders.FileSystemLoader object at 0x102df0190>, environment = <jinja2.environment.Environment object at 0x102df0130>, template = 'beginning.j2'

    def get_source(
        self, environment: "Environment", template: str
    ) -> t.Tuple[str, str, t.Callable[[], bool]]:
        pieces = split_template_path(template)
        for searchpath in self.searchpath:
            filename = os.path.join(searchpath, *pieces)
            f = open_if_exists(filename)
            if f is None:
                continue
            try:
>               contents = f.read().decode(self.encoding)
E               AttributeError: 'str' object has no attribute 'decode'

.tox/py38/lib/python3.8/site-packages/jinja2/loaders.py:201: AttributeError
```

I think I would just need to patch out the jinja2 functions I'm using in this case so that the patch is only patching the method I'm interested in.

**Comment by maishiroma** [comment]
# Update
- Addressed issues that came up with last reviews

@sarcasticadmin and @davido-nw if you want to review this again, feel free to!

File: scripts/gen_release_notes/tests/test_helper.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        "key": "RC/0.1",
+        "expected_value": "stg",
+    },
+    {"key": "https://github.com/some/repo", "expected_value": "url"},
+    {"key": "0.1.0", "expected_value": "prod"},
+    {"key": "someunknownvalue", "expected_value": None},
+]
+
+sample_regex_match = [
+    {"regex": r"^helloworld$", "context": "helloworld", "expected_result": True},
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/tests/test_helper.py:24]: Cool thanks for including the `expected_result` bool to ensure the test assertion is specific 🙂

File: scripts/gen_release_notes/setup.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    name="gen_release_notes",
+    description="Easily generate Markdown friendly release notes based on a Git repository",
+    author="NWI Engineering",
+    author_email="engineering@nebulaworks.com",
+    url="https://github.com/Nebulaworks/orion",
+    packages=find_packages(exclude=("tests")),
+    include_package_data=True,
+    install_requires=requirements,
+    cmdclass={"register": DisabledCommands, "upload": DisabledCommands},
+    entry_points={"console_scripts": ["genreleasenotes = generate:main"]},
```
Comment (sarcasticadmin) [inline scripts/gen_release_notes/setup.py:31]: Nice entry point 👍

**Review by sarcasticadmin — APPROVED** [review]
@maishiroma I really appreciate you taking and iterating on this after all the feedback! Looks ready to go to me :shipit:

**Comment by maishiroma** [comment]
# Tox Output

```
===================================================== test session starts ======================================================
platform darwin -- Python 3.8.3, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
cachedir: .tox/py38/.pytest_cache
collected 39 items

tests/test_file_writer.py ....                                                                                           [ 10%]
tests/test_git_parser.py .............                                                                                   [ 43%]
tests/test_helper.py ......................                                                                              [100%]

============================================================ PASSES ============================================================
___________________ test_write_to_file[user_input_cli_args0-user_input_dir_list0-user_input_changelog_obj0] ____________________
----------------------------------------------------- Captured stdout call -----------------------------------------------------
Successfully generated changelog file at sample.md!
___________________ test_write_to_file[user_input_cli_args0-user_input_dir_list1-user_input_changelog_obj0] ____________________
----------------------------------------------------- Captured stdout call -----------------------------------------------------
Successfully generated changelog file at sample.md!
___________________ test_write_to_file[user_input_cli_args1-user_input_dir_list0-user_input_changelog_obj0] ____________________
----------------------------------------------------- Captured stdout call -----------------------------------------------------
Successfully generated changelog file at sample.md!
___________________ test_write_to_file[user_input_cli_args1-user_input_dir_list1-user_input_changelog_obj0] ____________________
----------------------------------------------------- Captured stdout call -----------------------------------------------------
Successfully generated changelog file at sample.md!
______________________________________ test_gen_changelog_obj[user_input_changelog_obj0] _______________________________________
----------------------------------------------------- Captured stdout call -----------------------------------------------------
Currently parsing changes in repository, please wait. \
Successfully obtained changelog object!
====================================================== 39 passed in 0.43s ======================================================
___________________________________________________________ summary ____________________________________________________________
  py38: commands succeeded
  congratulations :)
```
